fish-rust/src/env/environment_impl.rs

```diff
@@ -0,0 +1,1225 @@
+            } else if query.function {
+                let mut node = self.base.locals.clone();
+                while node.next().is_some() {
+                    node = node.next().unwrap();
+                    if node.borrow().new_scope {
+                        break;
+                    }
+                }
+                result.status = remove_from_chain(&mut node, key);
+            } else {
+                panic!("Unknown scope");
+            }
+        } else if Self::remove_from_chain(&mut self.base.locals, key) {
+            // pass
+        } else if Self::remove_from_chain(&mut self.base.globals, key) {
+            result.global_modified = true;
+        } else if uvars()
+            .as_mut()
+            .expect("Should have non-null uvars in this function")
+            .remove(&key.to_ffi())
+        {
+            result.uvar_modified = true;
+        } else {
+            result.status = EnvStackSetResult::ENV_NOT_FOUND;
+        }
+        result
+    }
+
+    /// Push a new shadowing local scope.
+    pub fn push_shadowing(&mut self) {
+        // Propagate local exported variables.
+        let node = EnvNodeRef::new(true, None);
+        for cursor in self.base.locals.iter() {
+            for (key, val) in cursor.borrow().env.iter() {
+                if val.exports() {
+                    let mut node_ref = node.borrow_mut();
+                    // Do NOT overwrite existing values, since we go from inner scopes outwards.
+                    if node_ref.env.get(key).is_none() {
+                        node_ref.env.insert(key.clone(), val.clone());
+                    }
+                    node_ref.changed_exported();
+                }
+            }
+        }
+        let old_locals = mem::replace(&mut self.base.locals, node);
+        self.shadowed_locals.push(old_locals);
+    }
+
+    /// Push a new non-shadowing (inner) local scope.
+    pub fn push_nonshadowing(&mut self) {
+        self.base.locals = EnvNodeRef::new(false, Some(self.base.locals.clone()));
+    }
+
+    /// Pop the variable stack.
+    /// Return a list of the names of variables which were modified.
+    /// TODO: We return the variable names because we may need to dispatch changes,
+    /// for example if there is a local change to LC_ALL; but that is rare. How can
+    /// we avoid these copies in the common case?
+    pub fn pop(&mut self) -> Vec<WString> {
+        let popped: EnvNodeRef;
+        if let Some(next) = self.base.locals.next() {
+            popped = mem::replace(&mut self.base.locals, next);
+        } else {
+            // Exhausted the inner scopes, put back a shadowing scope.
+            if let Some(shadowed) = self.shadowed_locals.pop() {
+                popped = mem::replace(&mut self.base.locals, shadowed);
+            } else {
+                panic!("Attempt to pop last local scope")
+            }
+        }
+        let var_names = popped.borrow().env.keys().cloned().collect();
```

Comment:
nit: should avoid cloning the keys if possible

Reply:
This was actually a bug in the C++ implementation - we were potentially reading these keys without holding the lock.